File: mobile_app/lib/shared/models/address.dart
/// A saved business delivery location.
///
/// This is where goods are sent — not where the phone is. GPS may help fill the
/// fields in, but the typed values are always the source of truth.
class Address {
  const Address({
    required this.id,
    this.label = '',
    this.contactName = '',
    this.contactPhone = '',
    required this.line1,
    this.area = '',
    required this.city,
    required this.state,
    required this.pincode,
    this.country = 'India',
    this.landmark = '',
    this.latitude,
    this.longitude,
    this.isDefault = false,
  });

  final String id;
  final String label;
  final String contactName;
  final String contactPhone;
  final String line1;
  final String area;
  final String city;
  final String state;
  final String pincode;
  final String country;
  final String landmark;
  final double? latitude;
  final double? longitude;
  final bool isDefault;

  /// Shown as the row title. Falls back to the locality when unlabelled.
  String get displayLabel => label.trim().isNotEmpty ? label.trim() : (area.isNotEmpty ? area : city);

  /// "Patna, Bihar - 800001" — the compact second line in lists.
  String get shortLine {
    final place = [city, state].where((p) => p.trim().isNotEmpty).join(', ');
    return pincode.trim().isEmpty ? place : '$place - $pincode';
  }

  /// Full single-line rendering.
  String get formatted => [line1, area, city, state, pincode, country]
      .map((p) => p.trim())
      .where((p) => p.isNotEmpty)
      .join(', ');

  bool get hasCoordinates => latitude != null && longitude != null;

  factory Address.fromJson(Map<String, dynamic> json) => Address(
        id: json['id']?.toString() ?? '',
        label: json['label']?.toString() ?? '',
        contactName: json['contact_name']?.toString() ?? '',
        contactPhone: json['contact_phone']?.toString() ?? '',
        line1: json['line1']?.toString() ?? '',
        area: json['area']?.toString() ?? '',
        city: json['city']?.toString() ?? '',
        state: json['state']?.toString() ?? '',
        pincode: json['pincode']?.toString() ?? '',
        country: json['country']?.toString() ?? 'India',
        landmark: json['landmark']?.toString() ?? '',
        latitude: (json['latitude'] as num?)?.toDouble(),
        longitude: (json['longitude'] as num?)?.toDouble(),
        isDefault: json['is_default'] as bool? ?? false,
      );

  Map<String, dynamic> toRequestBody() => {
        'label': label,
        'contact_name': contactName,
        'contact_phone': contactPhone,
        'line1': line1,
        'area': area,
        'city': city,
        'state': state,
        'pincode': pincode,
        'country': country,
        'landmark': landmark,
        if (latitude != null) 'latitude': latitude,
        if (longitude != null) 'longitude': longitude,
      };

  Address copyWith({String? id, bool? isDefault}) => Address(
        id: id ?? this.id,
        label: label,
        contactName: contactName,
        contactPhone: contactPhone,
        line1: line1,
        area: area,
        city: city,
        state: state,
        pincode: pincode,
        country: country,
        landmark: landmark,
        latitude: latitude,
        longitude: longitude,
        isDefault: isDefault ?? this.isDefault,
      );
}

File: backend/src/models/Address.js
const mongoose = require('mongoose');

/**
 * A business delivery location. A business (User) owns many of these.
 *
 * Deliberately NOT the phone's GPS position: this is where goods are sent, which
 * is frequently nowhere near where the person placing the order is sitting.
 * latitude/longitude are optional metadata captured only when the user opts into
 * "use my current location", and are never authoritative over the typed fields.
 */
const addressSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },

    // Free-form name the business gives this location, e.g. "Main Warehouse".
    label: { type: String, default: '', trim: true },

    // Who receives the delivery here — may differ per location.
    contactName: { type: String, default: '', trim: true },
    contactPhone: { type: String, default: '', trim: true },

    line1: { type: String, required: true, trim: true }, // Address / Street
    area: { type: String, default: '', trim: true }, // Area / Locality
    city: { type: String, required: true, trim: true },
    state: { type: String, required: true, trim: true },
    pincode: { type: String, required: true, trim: true },
    country: { type: String, default: 'India', trim: true },
    landmark: { type: String, default: '', trim: true },

    // Only set when the user used GPS assistance; purely supplementary.
    latitude: { type: Number },
    longitude: { type: Number },

    isDefault: { type: Boolean, default: false },
  },
  { timestamps: true }
);

// Default address first, then most recently touched.
addressSchema.index({ userId: 1, isDefault: -1, updatedAt: -1 });

/** Single-line rendering used for list rows and legacy free-text fields. */
addressSchema.methods.formatted = function formatted() {
  return [this.line1, this.area, this.city, this.state, this.pincode, this.country]
    .map((part) => (part || '').trim())
    .filter(Boolean)
    .join(', ');
};

module.exports = mongoose.model('Address', addressSchema);
